fs/backgroundfetcher/background_fetcher.go

```diff
@@ -0,0 +1,152 @@
+	// The background fetcher will fetch a single span every `defaultFetchPeriod`.
+	defaultFetchPeriod = 500 * time.Millisecond
+)
+
+func WithMinFetchSize(size int64) Option {
+	return func(bf *BackgroundFetcher) error {
+		bf.minFetchSize = size
+		return nil
+	}
+}
+
+func WithSilencePeriod(period time.Duration) Option {
+	return func(bf *BackgroundFetcher) error {
+		bf.silencePeriod = period
+		return nil
+	}
+}
+
+func WithFetchPeriod(period time.Duration) Option {
+	return func(bf *BackgroundFetcher) error {
+		bf.fetchPeriod = period
+		return nil
+	}
+}
+
+// A backgroundFetcher is responsible for fetching spans from layers
+// in the background.
+type BackgroundFetcher struct {
+	minFetchSize  int64
+	silencePeriod time.Duration
+	fetchPeriod   time.Duration
+
+	rateLimiter *rate.Limiter
+
+	// All span managers are added to the channel and picked up in Run().
+	// If a span manager is still able to fetch, it is reinserted into the chanel.
+	waitQueue chan Resolver
+
+	closeChan chan struct{}
+
+	// The timer that the bg-fetcher will wait on until it is ready to (re)start
+	// background fetches.
+	timer *time.Timer
+}
+
+func NewBackgroundFetcher(opts ...Option) (*BackgroundFetcher, error) {
+	bf := newWithDefaults()
+	for _, o := range opts {
+		if err := o(bf); err != nil {
+			return nil, err
+		}
+	}
+	// Create a rate-limiter that will fetch every bf.fetchPeriod
+	// with a burst capacity of 1 (i.e., it will never invoke more than 1 bg-fetch
+	// within bf.fetchPeriod)
+	bf.rateLimiter = rate.NewLimiter(rate.Every(bf.fetchPeriod), 1)
+	return bf, nil
+}
+
+func newWithDefaults() *BackgroundFetcher {
+	return &BackgroundFetcher{
+		minFetchSize:  defaultMinFetchSize,
+		silencePeriod: defaultSilencePeriod,
+		fetchPeriod:   defaultFetchPeriod,
+		waitQueue:     make(chan Resolver),
+		closeChan:     make(chan struct{}),
+	}
+}
+
+// Add a new Resolver to be background fetched from.
+// Sends the resolver through the channel, which will be received in the Run() method.
+// Also resets the `timeTicker` so that background fetches are suspended until `silencePeriod`.
+func (bf *BackgroundFetcher) Add(resolver Resolver) {
+	bf.waitQueue <- resolver
+	if bf.timer == nil {
+		bf.timer = time.NewTimer(bf.silencePeriod)
+	} else {
+		bf.timer.Reset(bf.silencePeriod)
+	}
+}
+
+func (bf *BackgroundFetcher) Close() error {
+	bf.closeChan <- struct{}{}
+	return nil
+}
+
+func (bf *BackgroundFetcher) Run(ctx context.Context) error {
+	for {
+
+		// If any new span managers have come in, wait for silencePeriod.
+		if bf.timer != nil {
+			<-bf.timer.C
+		}
+
+		if err := bf.rateLimiter.Wait(ctx); err != nil {
+			return fmt.Errorf("background fetch: error while waiting for rate limiter: %w", err)
+		}
+
+		select {
+		case <-bf.closeChan:
+			return nil
+		case <-ctx.Done():
+			return nil
+		case lr := <-bf.waitQueue:
+			go func() {
+				more, err := lr.Resolve(ctx)
+				// if error, append it to the waitQueue to retry
```

Comment:
Would it retry that forever?

Reply:
Tbh, I wasn't sure what to do in this case? Retrying forever seems wasteful. Should we define a max error count?  And retry until it succeeds or error count reaches max error count? Something like this:

```
go func() {
   for i := 0; i < maxErrorCount; i++ {
       more, err := lr.Resolve(ctx)
       if err != nil {
           continue
        } else {
            break
       }
    }
}
```

Reply:
Looks good to me. We ideally should honor some HTTP status codes (e.g. don't retry in some cases) and headers (e.g. https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Retry-After), but we can start from just limiting the # of retries.

Reply:
We already use [retryablehttp](https://github.com/hashicorp/go-retryablehttp). https://github.com/awslabs/soci-snapshotter/blob/main/fs/remote/resolver.go#L184-L201

Reply:
What should the behavior be then? `retryablehttp` with retry a fixed number of times. After 'n' failed retries, should we just remove the layer from being background fetched?

Reply:
I'm inclined to do so. The availability gain by adding this retry wouldn't outweigh the risk of overloading a container registry. We should start from the specific scenarios that retryablehttp couldn't cover.